### PR TITLE
Get pair tweak

### PR DESCRIPTION
**Pull request by sfedrick**
Changes get pair to use the low weighting for one submission. If this probabilistically low info submission has enough information we compare it to a song that is close to it in devolving space.

### CODE REVIEW

File: devolving_music/lib/score_suite.py [hunk 1]
```diff
@@ -1,3 +1,4 @@
+from cmath import inf
```
Comment (cstuartroe) [inline devolving_music/lib/score_suite.py]: `cmath.inf` is for complex numbers. Use `math.inf`.

File: devolving_music/lib/song_scores.py [hunk 2]
```diff
@@ -29,18 +31,35 @@ def get_scores(self) -> "dict[int, ScoreSuite]":
 
         return self.song_score_dict
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]: Just noticed that we have a method starting with `get` that actually modifies the object's state. Let's call it `update_scores` and not have it return anything.

File: devolving_music/lib/song_scores.py [hunk 3]
```diff
@@ -29,18 +31,35 @@ def get_scores(self) -> "dict[int, ScoreSuite]":
...
-        # list
-        return self.song_score_dict.get(random.choice(key_list)).song_submission
+        return self.song_score_dict.get(
+            random.choice(key_list)).song_submission
+
+    def get_compare_submission_closest(self, score_suite_obj: "ScoreSuite", information_threshold : int, update_scores = False):
+        # information threshold is the threshhold at which we start to consider score_suites as being able to be close to score_suite_obj
+        # score_suites below this information threshhold are considered to be infinitely far away from score_suite_obj
+        if(update_scores):
+            self.get_scores()
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]: Why put that into this method? It seems unrelated, make it the job of the outside context to update scores if it wants to.

File: devolving_music/lib/song_scores.py [hunk 4]
```diff
@@ -29,18 +31,35 @@ def get_scores(self) -> "dict[int, ScoreSuite]":
...
+        score_dict = copy.deepcopy(self.song_score_dict)
+        target_id = score_suite_obj.song_submission.id
+        score_dict.pop(target_id)
+        score_list=list(score_dict.values())
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]:
```suggestion
        [
          score_suite
          for id, score_suite in self.song_score_dict.items()
          if id != score_suite_obj.id
        ]
```

A comprehension is much more readable and straightforward.

File: devolving_music/lib/song_scores.py [hunk 5]
```diff
@@ -29,18 +31,35 @@ def get_scores(self) -> "dict[int, ScoreSuite]":
...
     def get_compare_submission_random(self, submission_id) -> SongSubmission:
         key_list = list(self.song_score_dict.keys())
         key_list.remove(submission_id)
-        # once you have a critical number of comparisons then pull from quality
-        # list
-        return self.song_score_dict.get(random.choice(key_list)).song_submission
+        return self.song_score_dict.get(
+            random.choice(key_list)).song_submission
+
+    def get_compare_submission_closest(self, score_suite_obj: "ScoreSuite", information_threshold : int, update_scores = False):
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]:
```suggestion
    def get_compare_submission_closest(self, score_suite: "ScoreSuite", information_threshold : int, update_scores = False):
```

There's no reason to suffix a variable name with `_obj`. Everything is an object in Python.

File: devolving_music/lib/song_scores.py [hunk 6]
```diff
@@ -29,18 +31,35 @@ def get_scores(self) -> "dict[int, ScoreSuite]":
...
+        # information threshold is the threshhold at which we start to consider score_suites as being able to be close to score_suite_obj
+        # score_suites below this information threshhold are considered to be infinitely far away from score_suite_obj
+        if(update_scores):
+            self.get_scores()
+        score_dict = copy.deepcopy(self.song_score_dict)
+        target_id = score_suite_obj.song_submission.id
+        score_dict.pop(target_id)
+        score_list=list(score_dict.values())
+        closest_songs=SongScores.get_distance_sort(score_suite_obj, score_list, information_threshold)
+        return closest_songs
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]: I still don't like the pattern of creating a variable for the sole purpose of immediately returning it. If the variable name tells you something about what it is, the function name should give the same information anyway.

File: devolving_music/lib/song_scores.py [hunk 7]
```diff
@@ -110,42 +132,86 @@ def get_quality_list(self,
...
+
         """
-        score_suite_list.sort(key=lambda sub: sub.info_score)
+        score_suite_list = sorted(score_suite_list, key=lambda sub: sub.info_score)
         rightend = 0
-        init = score_suite_list[0].info_score
+        if (get_informed):
+            init = informed_threshold
+        else:
+            init = score_suite_list[0].info_score
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]: `init` is an uninformative variable name.

Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]:
```suggestion
        if not get_informed:
            informed_threshold = score_suite_list[0].info_score
```

Why not just keep using `informed_threshold`? It's a good name.

File: devolving_music/lib/score_suite.py (hunk 1 above)
Comment (cstuartroe) [reply devolving_music/lib/score_suite.py]: Why'd you resolve this?

File: devolving_music/lib/song_scores.py [hunk 8]
```diff
@@ -19,28 +19,52 @@ def __init__(self, event: Event):
...
-        # list
-        return self.song_score_dict.get(random.choice(key_list)).song_submission
+        return self.song_score_dict.get(
+            random.choice(key_list)).song_submission
+
+    def get_compare_submission_closest(self, score_suite_obj: "ScoreSuite", information_threshold : int, update_scores = False):
+        # information threshold is the threshhold at which we start to consider score_suites as being able to be close to score_suite_obj
+        # score_suites below this information threshhold are considered to be infinitely far away from score_suite_obj
+        if(update_scores):
+            self.update_scores()
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]: Still not sure why we'd be updating scores here rather than in some outer context. It doesn't seem related to this method.

I think there's also a good case to be made for taking it out of `get_scores_list` and `get_scores_dict`.

Actually, upon closer inspection, `update_scores` doesn't appear to be idempotent, so calling it all over the place might not only be wasteful but actively cause bugs. Why not just put it in the constructor?

File: devolving_music/lib/song_scores.py [hunk 9]
```diff
@@ -110,42 +137,86 @@ def get_quality_list(self,
...
+
+        for song_score in enumerate(info_list):
+            increase_ceiling = ceiling - info_list[song_score[0]].info_score
+            if (increase_ceiling < 0):
+                increase_ceiling = 0
+
+            if ((ceiling_check + increase_ceiling) >= select_num):
+                return song_score[1]
+            else:
+                ceiling_check += increase_ceiling
```
Comment (cstuartroe) [inline devolving_music/lib/song_scores.py]:
```suggestion
        for song_score in info_list:
            increase_ceiling = ceiling - song_score.info_score
            if (increase_ceiling < 0):
                increase_ceiling = 0

            if ((ceiling_check + increase_ceiling) >= select_num):
                return song_score
            else:
                ceiling_check += increase_ceiling
```

Needless use of `enumerate`.

File: devolving_music/lib/song_scores.py (hunk 4 above)
Comment (sfedrick) [reply devolving_music/lib/song_scores.py]: I disagree about this I normally don't find comprehensions more readable. It's very clear what each line does step by step. However, in the comprehension I have to think more about what is going on.  I'd agree if I had a for loop but pop is far more readable than a list comprehension.

File: devolving_music/lib/song_scores.py (hunk 5 above)
Comment (sfedrick) [reply devolving_music/lib/song_scores.py]: I know I like this because I won't mistakenly confuse score suite for the class score_suite. It makes it very clear it is the object of the class and not the class.

File: devolving_music/lib/song_scores.py (hunk 8 above)
Comment (sfedrick) [reply devolving_music/lib/song_scores.py]: It won't cause bugs because the comparison is never changing and that is what the update scores calculation is based on . But you're right about it being potentially wasteful I'll put it in the constructor.

File: devolving_music/lib/song_scores.py (hunk 9 above)
Comment (sfedrick) [reply devolving_music/lib/song_scores.py]: yeah you're right this is a remnant of when I was using the index for some part of the weighting.

File: devolving_music/lib/song_scores.py (hunk 6 above)
Comment (sfedrick) [reply devolving_music/lib/song_scores.py]: I still don't see why more information is bad. I personally like it. I think it makes things a lot easier to see than returning SongScores.get_distance_sort(score_suite_obj, score_list, information_threshold)

File: devolving_music/lib/score_suite.py (hunk 1 above)
Comment (sfedrick) [reply devolving_music/lib/score_suite.py]: I thought that was in song_scores my bad i fixed that one too
